paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc

```diff
@@ -156,11 +181,25 @@ class MatMulFactory {
 
   void Execute() {
     dnnl::stream stream(engine_);
-    matmul_prim_.execute(stream, {
-                                     {MKLDNN_ARG_SRC, x_mem_},
-                                     {MKLDNN_ARG_WEIGHTS, y_mem_},
-                                     {MKLDNN_ARG_DST, out_mem_},
-                                 });
+
+    auto offsets = offsets_;
+    unsigned bs = batch_size_;
+    void* x_ptr = x_mem_.get_data_handle();
+    void* y_ptr = y_mem_.get_data_handle();
+    void* out_ptr = out_mem_.get_data_handle();
+    for (unsigned i = 0; i < bs; i++) {
+      x_mem_.set_data_handle(x_ptr);
+      y_mem_.set_data_handle(y_ptr);
+      out_mem_.set_data_handle(out_ptr);
+      matmul_prim_.execute(stream, {
+                                       {MKLDNN_ARG_SRC, x_mem_},
+                                       {MKLDNN_ARG_WEIGHTS, y_mem_},
+                                       {MKLDNN_ARG_DST, out_mem_},
+                                   });
+      x_ptr = static_cast<char*>(x_ptr) + offsets.x_offset;
+      y_ptr = static_cast<char*>(y_ptr) + offsets.y_offset;
+      out_ptr = static_cast<char*>(out_ptr) + offsets.out_offset;
+    }
```

Comment:
Since this for loop isn't parallelised, wouldn't it be more efficient to calculate each offset on the go?
This way we avoid a cache miss on fetching offsets_ table which could potentially cost us more cycles than just increasing `x_ptr`, `y_ptr ` and `out_ptr ` by the respective products of each memory dimensions (Please see the [documentation](https://oneapi-src.github.io/oneDNN/structdnnl_1_1memory.html#afdd20764d58c0b517d5a31276672aeb8)).
An exemplary implementation could look like this:
```C++
    unsigned bs = batch_size_;
    void* x_ptr = x_mem_.get_data_handle();
    void* y_ptr = y_mem_.get_data_handle();
    void* out_ptr = out_mem_.get_data_handle();
    char* x_size = x_mem_.dims[1] * x_mem_.dims[2];
    char* y_size = y_mem_.dims[1] * y_mem_.dims[2];
    char* out_size = out_mem_.dims[1] * out_mem_.dims[2];
    for (unsigned i = 0; i < bs; i++) {
      x_mem_.set_data_handle(x_ptr);
      y_mem_.set_data_handle(y_ptr);
      out_mem_.set_data_handle(out_ptr);
      matmul_prim_.execute(stream, {
                                       {MKLDNN_ARG_SRC, x_mem_},
                                       {MKLDNN_ARG_WEIGHTS, y_mem_},
                                       {MKLDNN_ARG_DST, out_mem_},
                                   });
      x_ptr = static_cast<char*>(x_ptr) + x_size;
      y_ptr = static_cast<char*>(y_ptr) + y_size;
      out_ptr = static_cast<char*>(out_ptr) + out_size;
    }
```
This would also reduce cache trashing as we reduce the size of the class by removing the offsets_ field.

Reply:
it doesn't compile. 
` error: invalid use of ‘dnnl::memory::dims’`

Reply:
Anyway, matmul didn't work when we multiplied bathc size with 3rd dim so I guess batch 1 will also not work

Reply:
Checking this modification would add +1 / +2 weeks to development and testing

Reply:
> it doesn't compile.
> ` error: invalid use of ‘dnnl::memory::dims’`

It might be, I haven't compiled this but attempted to give a brief idea of what I mean, hence "An exemplary implementation could look like this:" disclaimer. 

> Anyway, matmul didn't work when we multiplied batch size with 3rd dim so I guess batch 1 will also not work

Hmm I don't quite understand how does this refer to the proposal? Nothing regarding batch size is changed here.

Reply:
@sfraczek and I have figured out that it's actually not possible to access the dimensions directly from memory primitive. 
Instead we'd need to fetch the memory descriptor and then from it the actual dimensions which are represented by a vector returned by value. 
Such copying of vectors and multiple levels of indirection would actually result in slower execution.